apis/oss/v1alpha1/oss_types.go

```diff
@@ -0,0 +1,84 @@
+/*
+
+ Copyright 2021 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package v1alpha1
+
+import (
+	runtimev1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +kubebuilder:object:root=true
+
+// OSSList contains a list of OSS
+type OSSList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []OSS `json:"items"`
+}
+
+// +kubebuilder:object:root=true
+
+// OSS is a managed resource that represents an OSS instance
+// +kubebuilder:printcolumn:name="READY",type="string",JSONPath=".status.conditions[?(@.type=='Ready')].status"
+// +kubebuilder:printcolumn:name="SYNCED",type="string",JSONPath=".status.conditions[?(@.type=='Synced')].status"
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster,categories={crossplane,managed,alibaba}
+type OSS struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   OSSSpec   `json:"spec,omitempty"`
+	Status OSSStatus `json:"status,omitempty"`
+}
+
+// OSSSpec defines the desired state of OSS
+type OSSSpec struct {
+	runtimev1.ResourceSpec `json:",inline"`
+	ForProvider            OSSParameters `json:"forProvider"`
+}
+
+// OSSStatus defines the observed state of OSS
+type OSSStatus struct {
+	runtimev1.ResourceStatus `json:",inline"`
+	AtProvider               OSSObservation `json:"atProvider,omitempty"`
+}
+
+// Bucket is the isolated place to store files
+type Bucket struct {
+	Name               string `json:"name"`
+	ACL                string `json:"acl,omitempty"`
+	StorageClass       string `json:"storageClass,omitempty"`
+	DataRedundancyType string `json:"dataRedundancyType,omitempty"`
+}
+
+// OSSParameters define the desired state of an OSS
+type OSSParameters struct {
+	Bucket Bucket `json:"bucket,omitempty"`
+}
+
+// OSSObservation is the representation of the current state that is observed.
+type OSSObservation struct {
+	BucketName       string `json:"bucketName,omitempty"`
+	ACL              string `json:"acl,omitempty"`
+	StorageClass     string `json:"storageClass,omitempty"`
+	RedundancyType   string `json:"redundancyType,omitempty"`
+	ExtranetEndpoint string `json:"extranetEndpoint,omitempty"`
+	IntranetEndpoint string `json:"intranetEndpoint,omitempty"`
+	Message          string `json:"message,omitempty"`
```

Comment:
does it mean a note?

Reply:
It's used to store the current status of OSS bucket object.

Especially, the message will store warning messages.

```
$ kubectl get oss --watch
NAME          READY   SYNCED   WARNING
example-oss   True    True     [Warning] StorageClass is not allowed to update after creation; [Warning] DataRedundancyType is not allowed to update after creation;
```

Reply:
cool, why not naming it status?

Reply:
I looked up `Condition` in crossplane-runtime api https://github.com/crossplane/crossplane-runtime/blob/master/apis/common/v1/condition.go#L60-L63. 
```
// A Condition that may apply to a resource.
type Condition struct {
	// Type of this condition. At most one of each condition type may apply to
	// a resource at any point in time.
	Type ConditionType `json:"type"`

	// Status of this condition; is it currently True, False, or Unknown?
	Status corev1.ConditionStatus `json:"status"`

	// LastTransitionTime is the last time this condition transitioned from one
	// status to another.
	LastTransitionTime metav1.Time `json:"lastTransitionTime"`

	// A Reason for this condition's last transition from one status to another.
	Reason ConditionReason `json:"reason"`

	// A Message containing details about this condition's last transition from
	// one status to another, if any.
	// +optional
	Message string `json:"message,omitempty"`
}
```

I need
1) create a new Condition as built-in Conditions `Creating`/`Deleting`/`Available`/`Unavailable`/`Error`/`Success` are not suitable.
2) and set Condition fields. But it's difficult to set fields `Status` and `Type` as they are unknown when the reconciler hit the warnings above.

Besides, I wanna users to quickly realize its settings is not proper and can have a quick fix or just ignore the warning.

So I choose this solution.